### PR TITLE
Calling finalize_state() when loading legacy h5 weights

### PR DESCRIPTION
**Pull request by Doch88**
Fixes: https://github.com/keras-team/keras/issues/20147

> So, I investigated it a bit and the problem is the keras.layers.Normalization layer, which is not finalized.
> This is because keras.layers.Normalization has two weights loaded from the file, adapted_mean and adapted_std, but these are not used directly when making a prediction.
> What is used directly are two variables, **which are not added as weights**, called mean and std, that are updated whenever finalize_state() is called.
> 
> And this is the problem. Theoretically, whenever loading weights from a file, finalize_state() should be called by load_own_variables().
> But load_own_variables() is called only with the new saving API, when the weights are saved in the format ".weights.h5", while the efficientnet weights have the legacy extension ".h5".
> 
> Therefore, weights are loaded, which means we have adapted_mean and adapted_std, but the actual mean and std are not loaded (since they're not weights) and instead, they're just initialized with the default values of build(), i.e. a vector of zeros and a vector of ones.

### CODE REVIEW

File: keras/src/legacy/saving/legacy_h5_format.py [hunk 1]
```diff
@@ -392,8 +396,59 @@ def load_weights_from_hdf5_group(f, model):
...
+                )
+                continue
+            raise ValueError(
+                f"Shape mismatch in {name}"
+                f"for weight {symbolic_weights[i].path}. "
+                f"Weight expects shape {expected_shape}. "
+                "Received saved weight "
+                f"with shape {received_shape}"
+            )
+        symbolic_weights[i].assign(weight_value)
```
Comment (fchollet) [inline keras/src/legacy/saving/legacy_h5_format.py:446]: Isn't this redundant with `load_own_variables`? If one is called, we shouldn't be calling the other.

Comment (Doch88) [reply keras/src/legacy/saving/legacy_h5_format.py:446]: I agree, but unfortunately I tried doing this instead:
```py
if hasattr(group, "load_own_variables"):
        variable_names = map(str, list(range(len(weight_values))))
        group.load_own_variables(dict(zip(variable_names, weight_values)))
    else:
        for i, weight_value in enumerate(weight_values):
            expected_shape = symbolic_weights[i].shape
            received_shape = weight_value.shape
            if expected_shape != received_shape:
                if skip_mismatch:
                    warnings.warn(
                        f"Skipping loading weights for {name}"
                        f"due to mismatch in shape for "
                        f"weight {symbolic_weights[i].path}. "
                        f"Weight expects shape {expected_shape}. "
                        "Received saved weight "
                        f"with shape {received_shape}",
                        stacklevel=2,
                    )
                    continue
                raise ValueError(
                    f"Shape mismatch in {name}"
                    f"for weight {symbolic_weights[i].path}. "
                    f"Weight expects shape {expected_shape}. "
                    "Received saved weight "
                    f"with shape {received_shape}"
                )
            symbolic_weights[i].assign(weight_value)
```
Which does either load_own_variables, if implemented, or the other method if not. Unfortunately, this breaks a test with layers having nested layers inside. 

Specifically this test: https://github.com/keras-team/keras/blob/3966392b8fdbdb9f369b0f01ffa319622bbaab2a/keras/src/saving/saving_lib_test.py#L962

The legacy h5 saving code puts the weights of the nested layers together with the weights of the layer that contains it. In the mentioned test, for example, `weight_values` contains 5 weights: the three weights of MyDense and the two weights of the nested layer. The base implementation of load_own_variables(), instead, asks for only the three weights of the top layer.

So, as I see it, either we:
- Find a way inside _set_weights to differentiate which weight belongs to the nested layers.
- Change the implementation of the saving method for legacy h5, but this could break retrocompatibility.
- Find another way to call the "post-processing" steps of load_own_variables(), e.g. the finalize_step() function needed by keras.layers.Normalization

Do you have any idea or suggestion?

Comment (fchollet) [reply keras/src/legacy/saving/legacy_h5_format.py:446]: I'd be concerned that the current method will add significant overhead to weights loading (it basically doubles total time taken since we're assigning variables twice).

A simple alternative may be to always call `finalize_state()` here after setting variables, if the method is defined and if there were weights that were changed. Basically `if hasattr(layer, "finalize_state") and symbolic_weights: layer.finalize_state()`, after all weight assignments are done. I think this should generalize well.

Comment (Doch88) [reply keras/src/legacy/saving/legacy_h5_format.py:446]: Fair enough, good point!
I was thinking that perhaps there are layers that perform other kinds of operations on load_own_variables() that may bring problems if they're not called. 
But as of now, I haven't noticed any other layer having this problem. 
And if it happens in the future, we just need to implement finalize_state()

Comment (Doch88) [reply keras/src/legacy/saving/legacy_h5_format.py:446]: Implemented the suggested change, tests seem to succeed (including the newly added one)

**Review by fchollet — APPROVED** [review]
LGTM, thank you!
